File: bubble-sort/bubble-sort.js
const generateArray = (size) => {
    const arr = []
    for (let i = 1; i <= size; i++) {
        const index = Math.floor(Math.random() * i)
        arr.splice(index, 0, i)
    }
    return arr
}

const bubbleSort = (arr) => {
    const res = arr.slice(0)
    for (let i = 0; i < res.length; i++) {
        for (let j = res.length - 1; j > i; j--) {
            if (res[j] < res[j - 1]) {
                const tmp = res[j]
                res[j] = res[j - 1]
                res[j - 1] = tmp
            }
        }
    }
    return res
}

const main = () =>
{
    const size = 30000

    // Generate Array
    const genStart = new Date().getTime()
    const arr = generateArray(size)
    const genEnd = new Date().getTime()

    // Sort array
    const sortStart = new Date().getTime()
    const res = bubbleSort(arr)
    const sortEnd = new Date().getTime()

    // Eval sorted array with an expected result
    const expected = []
    for (let i = 1; i <= size; i++) expected.push(i)
    areEqual = expected.toString() == res.toString()

    // Output the results
    console.log(`1. Time to generate array: ${genEnd - genStart} ms`)
    if (! areEqual) {
        console.log("2. The result array is NOT sorted as expected")
    } else {
        console.log("2. The result array is sorted as expected")
        console.log(`3. Time to bubble sort the array: ${sortEnd - sortStart} ms`)
    }
}

main()

File: python/bubble-sort.py
from math import floor
from time import time
from array import array

from arrays import (generate_shuffled_array,
                    is_sorted_as_expected,
                    generate_expected_array,
                    copy_array)
from time_utils import get_time_ms

def __bubble_sort(arr):
    for i in range(0, len(arr)):
        for j in range(len(arr) - 1, i, -1):
            if arr[j] < arr[j - 1]:
                # Swap
                arr[j], arr[j- 1] = arr[j - 1], arr[j]
    return arr

def bubble_sort(arr):
    copy = copy_array(arr)
    start = time()
    res = __bubble_sort(copy)
    end = time()
    sort_time = get_time_ms(start, end)
    return res, sort_time


def main():
    size = 40000

    # Generate shuffled array
    arr, gen_time = generate_shuffled_array(size)

    # Sort shuffled array
    res, sort_time = bubble_sort(arr)

    # Eval sorted array with an expected result
    expected = generate_expected_array(size)

    # Output the results
    print(f"1. Time to generate shuffled array of size {size}: {gen_time} ms")
    if not is_sorted_as_expected:
        print("2. The result array is NOT sorted as expected")
    else:
        print("2. The result array is sorted as expected")
        print(f"3. Time to bubble sort the shuffled array: {sort_time} ms")


main()
